### DISCUSSION TITLE
BMR - Logic of Rotation

### DISCUSSION THREAD

**Discussion by canhlinh** [icon-project/btp#13] [Q&A] [answered]
Greeting everyone,

We're in charge of developing BMR and at the moment we have a question regarding the rotation code of our BTP project.

- Here is the function [rorate](https://github.com/icon-project/btp/blob/master/cmd/btpsimple/chain/chain.go#L454)
- Here is the function [skippable](https://github.com/icon-project/btp/blob/master/cmd/btpsimple/chain/chain.go#L421)

From my view, the BMR and BMC are using the same strategy to check whether the next relay message is able to relay by a BMR. However I'm not sure the logics behind.

It would be great if you guys can help us clarify the below questions:
- How does the logic of rotation work?
- What is the purpose of the function **skippable** here? 
- What are the differences in terms of logic between rotation calculation  [here](https://github.com/icon-project/btp/blob/master/cmd/btpsimple/chain/chain.go#L428) and [here](https://github.com/icon-project/btp/blob/master/cmd/btpsimple/chain/chain.go#L469)? Why are they not the same?

Sincerely,

**Comment by jspark-icon** [comment, marked as answer]
- How does the logic of rotation work?
A: To rotation, basically using Round-Robin with list of registered BMR in BMC.
    Term of relayble of each BMR is configured by BMC.setLink method, and unit of term is based on height of block. also, it is related to 'max_aggregation' of BMC.
    When becoming turn, BMR could delay to relay 'BlockUpdate' until end of term. but BMR should relay 'delayed - BlockUpdates' and 'ReceiptProofs' immediately at BTP-Message creation for reduce latency. in this case, if BMR doesn't relay on time ('delay_limit'), then rotate to next.

- What is the purpose of the function skippable here?
A: check rotation when BMR doesn't relay immediately at BTP-Message creation

- What are the differences in terms of logic between rotation calculation here and here? Why are they not the same?
A: rotate calculation as below
in '_roate()' : bs.RotateHeight <--- rotateTerm / time1 ---> monitorHeight
in '_skippable()' : bs.RotateHeight <--- rotateTerm / time1 ---> rm.heightOfDst <--- delayLimit / time2 ---> monitorHeight
